decentraland/unity-explorer
# Fix: Reference Counting Cache refactoring

Fixes #2126

- Unified cache with ref counting capabilities for all suitable streamables:
   - Removed faulty logic that does not take into account real referencing count but relies on timestamp only
   - Introduced `RefCountStreamableCacheBase` for:
      - `Profiles`
      - `Asset Bundles`
      - `Audio Clips`
      - `Nft Shape`
      - `Textures`
   - It follows FIFO principle on clean-up: the last used asset will be unloaded first
   - Restricted to the only way to dereference via `StreamableRefCountData<TAsset>`, it is the base class for all cached assets with ref counting
- Introduced reference counting for Textures:
   - Before we didn't care about it so there were many places required attention:
      - `Materials` already were releasing properly: they required minimum adjustments
      - Introduced proper handling of:
         - Profile Pictures
         - NFT Shapes
         - Avatar Attachment Thumbnail (Emotes and Wearables)
         - UI Background
         - Map Pins

## How to QA
Following this pattern:
- Find the scene with the corresponding asset/assets
- Press `Memory.FULL` in `Debug Menu`
- The given asset(-s) should remain in tact
- Press `Memory.NORMAL`
- Leave the scene (till it's unloaded or changed to LOD)
All the affected assets should be cheched:
- NFT Shapes
- UI Background
- Map Pins
- Textures
- Asset Bundles

For Avatar Attachments check that thumbnails do not disappear (backpack, passport) after setting `Memory.FULL`
